### ISSUE TITLE
AOSP related issues

### ISSUE THREAD

**Issue by kirklandsign** [pytorch/executorch#6470]
Using this issue as the main thread for discussing AOSP related issues.

Original discussion: https://github.com/pytorch/executorch/commit/51cef8c817d5d0aca8a3c80db6b0ffc8de639ae2#diff-757cd72756ef920fd2357653c76c087315969557e257a0fc8532ed2c6dca1649R105

Known issues:
1. java test coverage <- WIP
2. [high priority] use fd in data loader like https://cs.android.com/android/platform/superproject/main/+/main:external/federated-compute/fcp/tensorflow/file_descriptor_filesystem.cc
3. java stability
4. question about HybridData, do we really need it? 
[executorch/extension/android/src/main/java/org/pytorch/executorch/Tensor.java](https://github.com/pytorch/executorch/blob/6e871c3b617fb3ff72dc8ad155e2f04ab441be3f/extension/android/src/main/java/org/pytorch/executorch/Tensor.java#L342) <- Yes
5. when port executorch v0.3.0, got error "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository." It means this release is not cut from main branch and qiaoli31 can't port into aosp. Is there a way to fix this tag and make it cut from main branch. In addition, could we make sure we cut release from main branch in the future
6. [high priority] could we expose a serializable format of EValue and Tensor class? The usage is partner can use this class to create Object[] for inferenceInput and ODP can use it to convert byte[] back to these data class when run executorch. We have deadline mid Nov including code complete and API review, so this is request is a little bit urgent. Let me know if this request makes sense or not.
https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/OnDevicePersonalization/framework/java/android/adservices/ondevicepersonalization/InferenceInput.java;l=56?q=inferenceinput

**Comment by qiaoli31** [comment]
6. [high priority] could we expose a serializable format of EValue and Tensor class? The usage is partner can use this class to create Object[] for inferenceInput and ODP can use it to convert byte[] back to these data class when run executorch. We have deadline mid Nov including code complete and API review, so this is request is a little bit urgent. Let me know if this request makes sense or not.  
https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/OnDevicePersonalization/framework/java/android/adservices/ondevicepersonalization/InferenceInput.java;l=56?q=inferenceinput

**Comment by qiaoli31** [comment]
we need the change in main branch not only AOSP.  yes, we need this class to be stable because it will be used by both partner and ODP. 

btw for request 1-5 as well, we need it in main branch, otherwise we need resolve conflict or manually edit every time because we are not allowed to port "random" branch. This aosp branch is only for testing purpose for us as well.

**Comment by kirklandsign** [comment]
> 6. [high priority] could we expose a serializable format of EValue and Tensor class? The usage is partner can use this class to create Object[] for inferenceInput and ODP can use it to convert byte[] back to these data class when run executorch. We have deadline mid Nov including code complete and API review, so this is request is a little bit urgent. Let me know if this request makes sense or not.
>    https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/OnDevicePersonalization/framework/java/android/adservices/ondevicepersonalization/InferenceInput.java;l=56?q=inferenceinput

Oh is this for Java layer only?

**Comment by qiaoli31** [comment]
EValue is quite straightforward as long as Tensor is serializable, all other types are  primitive types.
https://github.com/pytorch/executorch/blob/eea172de0d4c16ff2a831604c49efb9f47cc8df6/extension/android/src/main/java/org/pytorch/executorch/EValue.java

**Comment by qiaoli31** [comment]
Java class only. More specifically is these two classes. 
https://github.com/pytorch/executorch/blob/eea172de0d4c16ff2a831604c49efb9f47cc8df6/extension/android/src/main/java/org/pytorch/executorch/EValue.java
https://github.com/pytorch/executorch/blob/eea172de0d4c16ff2a831604c49efb9f47cc8df6/extension/android/src/main/java/org/pytorch/executorch/Tensor.java

**Comment by kirklandsign** [comment]
Sounds good. Basically we put a header for type and for tensor, the sizes and dim order. I will try to define one and ask our team for a review.

**Comment by qiaoli31** [comment]
Do we have ETA for item (6) which is the highest priority?

**Comment by qiaoli31** [comment]
sounds good, thank you!

**Comment by JacobSzwejbka** [comment]
>[high priority] could we expose a serializable format of EValue and Tensor class? The usage is partner can use this class to create Object[] for inferenceInput and ODP can use it to convert byte[] back to these data class when run executorch.

I am confused. When are EValues being sent over the wire? I thought we would be receiving a tf.example which contains a Dict[str,Tensor] and then have to have some code to convert those to proper ExecuTorch inputs? What exactly will the tf.example contain and what specific types are involved here?

**Comment by qiaoli31** [comment]
we have separate API interface for inference and training. tf.example is for training we can talk later. 

ModelManager is the inference API which takes Object[] as input data and generate Map<Integer, Object> as output.
https://cs.android.com/android/platform/superproject/+/android14-qpr3-release:packages/modules/OnDevicePersonalization/framework/java/android/adservices/ondevicepersonalization/ModelManager.java;l=64?q=modelmanager
sample code: 
https://cs.android.com/android/platform/superproject/+/android14-qpr3-release:packages/modules/OnDevicePersonalization/samples/OdpSampleNetwork/src/com/example/odpsamplenetwork/SampleHandler.java;l=765

Option 1:
partner: convert EValue[] (serialized format) -> Object[]
ODP: receive Object[] -> byte buffer -> convert it back to EValue when call executorch. 

We need a serialized format EValue to across IPC and don't want to own any custom conversion logic. IMO we should reuse this for training as well. 

FYI pytorch is ported into AOSP:
https://cs.android.com/android/platform/superproject/main/+/main:external/pytorch/

**Comment by kirklandsign** [comment]
> FYI pytorch is ported into AOSP:
https://cs.android.com/android/platform/superproject/main/+/main:external/pytorch/

Thank you! Since we have pytorch/pytorch and pytorch/executorch, would it be easier if the codepath is in external/pytorch/pytorch?

**Comment by kirklandsign** [comment]
@qiaoli31 Thanks for the explanation. I have some basic ideas about serialization

**Comment by davidlin54** [comment]
Added FD data loader in https://github.com/pytorch/executorch/tree/main/extension/data_loader/file_descriptor_data_loader.cpp#L95, unit tests in https://github.com/pytorch/executorch/tree/main/extension/android/src/test/java/org/pytorch/executorch. Still working on the integration tests for the Module and LlamaModule.

cc: @qiaoli31

**Comment by JacobSzwejbka** [comment]
Any other high priority blockers @qiaoli31 ?
